cli/sdo/voucher.go

```diff
@@ -151,6 +151,126 @@ type ImportResponse struct {
+func getVouchers (org, userCreds, apiMsg string, voucher string) ([]byte, string) {
+	msgPrinter := i18n.GetMessagePrinter()
+	cliutils.Verbose(msgPrinter.Sprintf("Listing imported SDO vouchers."))
+
+	// setup HTTP parameters and URL
+	var respBodyBytes []byte
+	var requestBodyBytes []byte
+	var sdoURL string
+	url := cliutils.GetSdoSvcUrl()
+	
+	if voucher == "" {
+		sdoURL = url+"/vouchers"
+	} else {
+		sdoURL = url+"/vouchers"+"/"+voucher 
+	}
+
+	creds := cliutils.OrgAndCreds(org, userCreds)
+	method := http.MethodGet
+	apiMsg = method + " " + sdoURL
+	httpClient := cliutils.GetHTTPClient(config.HTTPRequestTimeoutS)
+
+	resp := cliutils.InvokeRestApi(httpClient, method, sdoURL, creds, requestBodyBytes, "SDO Owner Service", apiMsg)
+	defer resp.Body.Close()
+	httpCode := resp.StatusCode
+	cliutils.Verbose(msgPrinter.Sprintf("HTTP code: %d", httpCode))
+
+	respBodyBytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		cliutils.Fatal(cliutils.HTTP_ERROR, msgPrinter.Sprintf("failed to read exchange body response from %s: %v", apiMsg, err))
+	}
+	if httpCode != 200 {
+		cliutils.Fatal(cliutils.HTTP_ERROR, msgPrinter.Sprintf("bad HTTP code %d from %s: %s", httpCode, apiMsg, string(respBodyBytes)))
+	}
+	if err != nil {
+		cliutils.Fatal(cliutils.HTTP_ERROR, msgPrinter.Sprintf("json unmarshalling HTTP response '%s' from %s: %v", string(respBodyBytes), apiMsg, err))
+	}
+
+	return respBodyBytes, apiMsg
+}
+
+// called when the -l flag is used to list the full voucher
+func listFullVoucher (respBodyBytes []byte, apiMsg string) []byte {
+	msgPrinter := i18n.GetMessagePrinter()
+	cliutils.Verbose(msgPrinter.Sprintf("Listing imported SDO vouchers."))
+
+	var output interface{}
+	err := json.Unmarshal(respBodyBytes, &output)
+	if err != nil {
+		cliutils.Fatal(cliutils.HTTP_ERROR, msgPrinter.Sprintf("json unmarshalling HTTP response '%s' from %s: %v", string(respBodyBytes), apiMsg, err))
+	}
+
+	jsonBytes, err := json.MarshalIndent(output, "", cliutils.JSON_INDENT)
+	if err != nil {
+		cliutils.Fatal(cliutils.JSON_PARSING_ERROR, msgPrinter.Sprintf("failed to marshal 'hzn exchange service list' output: %v", err))
+	}
+
+	return jsonBytes
+}
+
+// list the all the uploaded SDO vouchers, or a single voucher
+func VoucherList(org, userCreds, voucher string, namesOnly bool) {
+	msgPrinter := i18n.GetMessagePrinter()
+	cliutils.Verbose(msgPrinter.Sprintf("Listing imported SDO vouchers."))
+
+	// call the ocs-api to get the uploaded vouchers
+	var respBodyBytes []byte
+	var apiMsg string
+	respBodyBytes, apiMsg = getVouchers(org, userCreds, apiMsg, voucher)
+
+	// no voucher was specified, list all uploaded vouchers
+	if voucher == "" {
+		output := []string{}
+		err := json.Unmarshal(respBodyBytes, &output)
+		if err != nil {
+			cliutils.Fatal(cliutils.HTTP_ERROR, msgPrinter.Sprintf("json unmarshalling HTTP response '%s' from %s: %v", string(respBodyBytes), apiMsg, err))
+		}
+
+		jsonBytes, err := json.MarshalIndent(output, "", cliutils.JSON_INDENT)
```

Comment:
Did you test that the response looks correct when there are no vouchers? Sometimes we've been bitten by that with the hzn sub-cmds.

Can't remember if i told you how to get the sdo-owner-services container to forget about the vouchers that have been imported: when the container is not running: `docker volume rm sdo-ocs-db`

Reply:
I'll double check!

Reply:
Okay yeah, looked good 
```
Troys-MacBook-Pro:anax troyfine$ cli/hzn voucher list 
[]
```